File: Lógica de Programación/#02 FUNCIONES Y ALCANCE/Python/Hotchkyssdev.py
'''
FUNCIONES Y ALCANCE
'''

#Funciones definidas por el usuario

#Simple
def saludar():
    print("Hola Python.")

saludar()

#Con Retorno
def devolver_saludo():
    return "Hola Python."

print(devolver_saludo())
print("")

#Con un Argumento
def args_saludo(saludo, nombre):
    print(f"{saludo}, {nombre}!")

args_saludo("Hola", "Tomás")
args_saludo(nombre = "Tomás", saludo = "Hola")

print("")

#Con un argumento predeterminado
def default_arg_saludo(nombre = "Python"):
    print(f"Hola, {nombre}!")

default_arg_saludo("Tomás")
default_arg_saludo()
print("")

#Con Argumentos y Retorno
def retorno_args_saludo(saludo, nombre):
    return f"{saludo}, {nombre}!"

print(retorno_args_saludo("Hola", "Tomás")) 
print("")

#Con retorno de varios valores
def multiple_retorno_saludo():
    return "Hola", "Python"

saludo, nombre = multiple_retorno_saludo()
print(saludo)
print(nombre)
print("")

#Con un número variable de argumentos
def variable_args_saludo(*nombres):
    for nombre in nombres:
        print(f"Hola, {nombre}!")

variable_args_saludo("Python", "Gersom", "Hotchkyss", "Programadores")
print("")

#Con un número variable de argumentos con palabra clave
def variable_llave_args_saludo(**nombres):
    for llave, valor in nombres.items():
        print(f"Hola, {valor} ({llave})!")

variable_llave_args_saludo(
    lenguaje = "Python", 
    nombre = "Gersom", 
    alias = "HotchkyssDev", 
    edad = 18
)
print("")

'''
FUNCIONES DENTRO DE FUNCIONES
'''

def funcion_externa():
    def funcion_interna():
        print("Funcion interna: Hola, Python!")
    funcion_interna()

funcion_externa()
print("")

'''
FUNCIONES DEL LENGUAJE (built-in)
'''

print(len("HotchkyssDev"))
print(type(18))
print("HotchkyssDev".upper())
print("HOTCHKYSSDEV".lower())
print("")

'''
VARIABLE LOCAL Y GLOBAL
'''

variable_global = "Python"
print(variable_global)

def hola_python():
    variable_local = "Hola"
    print(f"{variable_local}, {variable_global}")

print(variable_global)
#print(variable_local) No se puede acceder desde fuera de la funcion

hola_python()
print("")

'''
DIFICULTAD EXTRA: FIZZ BUZZ

Crea una función que reciba dos parámetros de tipo cadena de texto y retorne un número.
 * - La función imprime todos los números del 1 al 100. Teniendo en cuenta que:
 *   - Si el número es múltiplo de 3, muestra la cadena de texto del primer parámetro.
 *   - Si el número es múltiplo de 5, muestra la cadena de texto del segundo parámetro.
 *   - Si el número es múltiplo de 3 y de 5, muestra las dos cadenas de texto concatenadas.
 *   - La función retorna el número de veces que se ha impreso el número en lugar de los textos.
'''

def numeros(str1, str2)-> int:
    contador = 0
    for number in range(1,101):
        if number % 3 == 0 and number % 5 == 0:
            print(str1, str2)
        elif number % 3 == 0:
            print(str1)
        elif number % 5 == 0:
            print(str2)
        else:
            print(number)
            contador = contador + 1
    return contador

print("La cantidad de números que devuelve la función es: ", numeros("Fizz", "Buzz"))
print("")

File: Lógica de Programación/#02 FUNCIONES Y ALCANCE/C/Hotchkyssdev.c
//FUNCIONES Y ALCANCE
#include <stdio.h>
#include <stdarg.h>

//Funciones definidas por el usuario
//Simple
void funcion1() {
    printf("Hola desde la funcion 1\n");
}

//Con retorno
int funcion2() {
    return 42;
}

//Con un argumento
void funcion3(int x) {
    printf("El valor de la funcion3 es: %d\n", x);
}

//Con Argumentos y Retorno
int funcion4(int a, int b) {
    return a + b;
}

//Con retorno de varios valores
void funcion5(int a, int b, int *suma, int *producto) {
    *suma = a + b;
    *producto = a * b;
}

//Con un número variable de argumentos
int funcion6(int count, ...) {
    va_list args;
    int sum = 0;
    va_start(args, count);
    for (int i = 0; i < count; i++) {
        sum += va_arg(args, int);
    }
    va_end(args);
    return sum;
}

//Con un número variable de argumentos con palabra clave
int funcion7(const char *operation, int count, ...) {
    va_list args;
    int result = (operation[0] == 's') ? 0 : 1; 
    va_start(args, count);
    for (int i = 0; i < count; i++) {
        int value = va_arg(args, int);
        if (operation[0] == 's') {
            result += value;
        } else if (operation[0] == 'p') {
            result *= value;
        }
    }
    va_end(args);
    return result;
}

/*DIFICULTAD EXTRA: FIZZ BUZZ

Crea una función que reciba dos parámetros de tipo cadena de texto y retorne un número.
 * - La función imprime todos los números del 1 al 100. Teniendo en cuenta que:
 *   - Si el número es múltiplo de 3, muestra la cadena de texto del primer parámetro.
 *   - Si el número es múltiplo de 5, muestra la cadena de texto del segundo parámetro.
 *   - Si el número es múltiplo de 3 y de 5, muestra las dos cadenas de texto concatenadas.
 *   - La función retorna el número de veces que se ha impreso el número en lugar de los textos.
*/

int fizzBuzz(const char *string1, const char *string2) {
    int contador = 0;
    int i = 1;
    for (i = 1; i <= 100; i++) {
        if (i % 3 == 0  && i % 5 == 0) {
            printf("%s%s\n", string1, string2);
        } else if (i % 3 == 0) {
            printf("%s\n", string1);
        } else if (i % 5 == 0) {
            printf("%s\n", string2);
        } else {
            printf("%d\n", i);
            contador++;
        }
    }

    return contador;
}

int main() {
    funcion1();
    printf("Retorno de funcion2: %d\n", funcion2());
    funcion3(10);
    printf("Retorno de funcion4: %d\n", funcion4(5, 7));
    
    int suma, producto;
    funcion5(3, 4, &suma, &producto);
    printf("La funcion 5 devuelve Suma: %d, Producto: %d\n", suma, producto);
    
    printf("La funcion 6 devuelve Suma variable (1,2,3): %d\n", funcion6(3, 1, 2, 3));
    printf("La funcion 7 devuelve Producto variable (2,3,4): %d\n\n", funcion7("p", 3, 2, 3, 4));
    
    int contador = fizzBuzz("Fizz", "Buzz");
    printf("\nNumeros impresos: %d\n", contador);

    return 0;
}
